### ISSUE TITLE
Package not restored from S3 based repository

### ISSUE THREAD

**Issue by vspinu** [rstudio/renv#848]
Renv fails to install an older version of the package from an S3 based repo:

```
{
  "R": {
    "Version": "4.1.1",
    "Repositories": [
      {
        "Name": "CRAN",
        "URL": "https://cloud.r-project.org"
      },
      {
        "Name": "ARROW",
        "URL": "https://arrow-r-nightly.s3.amazonaws.com"
      }
    ]
  },
  "Packages": {
    "renv": {
      "Package": "renv",
      "Version": "0.14.0",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "30e5eba91b67f7f4d75d31de14bbfbdc"
    },
    "arrow": {
      "Package": "arrow",
      "Version": "5.0.0.20211025",
      "Source": "Repository",
      "Repository": "ARROW",
      "Hash": "7b5cbd177248782f8b21b2494e9b97cf"
    }
  }
}
```

Both `renv:restore` and direct installation `renv::install("arrow@5.0.0.20211025")` fail to install the package, even though the package is actually there (https://arrow-r-nightly.s3.amazonaws.com/src/contrib/arrow_5.0.0.20211025.tar.gz)

On the other hand both `restore` and `install` work with a more recent version `6.0.0.20211031`. 

I susspect this has to do with s3 listing limit of 1000 items. If that's the case then, maybe renv could try to fetch directly without fetching the listin?

**Comment by kevinushey** [comment]
What error do you see when you try this?

**Comment by vspinu** [comment]
Retrieval failure:

```
> renv::install("arrow@5.0.0.20211025")
Error: failed to retrieve package 'arrow'
Traceback (most recent calls last):
7: renv::install("arrow@5.0.0.20211025")
6: retrieve(names(remotes))
5: handler(package, renv_retrieve_impl(package))
4: renv_retrieve_impl(package)
3: renv_retrieve_repos(record)
2: stopf("failed to retrieve package '%s'", record$Package)
1: stop(sprintf(fmt, ...), call. = call.)
```

**Comment by kevinushey** [comment]
As far as I can tell, that version of `arrow` just doesn't exist in that repository.

```
> db <- renv:::renv_available_packages_impl(
+   type = "source",
+   repos = "https://arrow-r-nightly.s3.amazonaws.com"
+ )
* Querying repositories for available source packages ... Done!
> 
> str(db)
List of 1
 $ :'data.frame':	1 obs. of  16 variables:
  ..$ Package              : chr "arrow"
  ..$ Version              : chr "6.0.0.20211101"
  ..$ Priority             : chr NA
  ..$ Depends              : chr "R (>= 3.3)"
  ..$ Imports              : chr "assertthat, bit64 (>= 0.9-7), methods, purrr, R6, rlang,\nstats, tidyselect, utils, vctrs"
  ..$ LinkingTo            : chr NA
  ..$ Suggests             : chr "DBI, dbplyr, decor, distro, dplyr, duckdb (>= 0.2.8), hms,\nknitr, lubridate, pkgload, reticulate, rmarkdown, s"| __truncated__
  ..$ Enhances             : chr NA
  ..$ License              : chr "Apache License (>= 2.0)"
  ..$ License_is_FOSS      : chr NA
  ..$ License_restricts_use: chr NA
  ..$ OS_type              : chr NA
  ..$ Archs                : chr NA
  ..$ MD5sum               : chr "5b181a27e69c7aa5ed9377e81843fd87"
  ..$ NeedsCompilation     : chr "yes"
  ..$ Repository           : chr "https://arrow-r-nightly.s3.amazonaws.com/src/contrib"
```

```
> url <- "https://arrow-r-nightly.s3.amazonaws.com/src/contrib/Archive/arrow/arrow_5.0.0.20211025.tar.gz"
> renv:::renv_download_available(url)
[1] FALSE
```

Because you're asking `renv::restore()` to restore a package version that doesn't exist, the restore fails. You're probably hoping that `renv` could auto-update to whatever the latest version happens to be; if you want that behavior, you should prefer just using `renv::install()`.

**Comment by vspinu** [comment]
The URL is actually  https://arrow-r-nightly.s3.amazonaws.com/src/contrib/arrow_5.0.0.20211025.tar.gz  It's the same url format as reported by `install.packages`:

```
> install.packages("arrow", repos = "https://arrow-r-nightly.s3.amazonaws.com")
Installing package into ‘/home/vspinu/smava/proj/xray/base/renv/library/R-4.0/x86_64-pc-linux-gnu’
(as ‘lib’ is unspecified)
trying URL 'https://arrow-r-nightly.s3.amazonaws.com/src/contrib/arrow_6.0.0.20211102.tar.gz'
Content type 'binary/octet-stream' length 4495447 bytes (4.3 MB)
```

**Comment by vspinu** [comment]
What versions are you referring to? 

I really mean `5.0.0.20211025` which is there according to 
 https://arrow-r-nightly.s3.amazonaws.com/src/contrib/arrow_5.0.0.20211025.tar.gz

the path is different from the link you have tried, namely: 
https://arrow-r-nightly.s3.amazonaws.com/src/contrib/Archive/arrow/arrow_5.0.0.20211025.tar.gz

My link conforms to the link produced by `install.packages` for the most recent version of arrow (version of which is irrelevant):
https://arrow-r-nightly.s3.amazonaws.com/src/contrib/arrow_6.0.0.20211102.tar.gz

**Comment by kevinushey** [comment]
Thanks; I see what you mean now. Sorry for the confusion.

In that case, I believe the issue here is that the `PACKAGES.rds` file does not actually declare that this version of `arrow` is available:

```
url <- "https://arrow-r-nightly.s3.amazonaws.com/src/contrib/PACKAGES.rds"
destfile <- tempfile(fileext = ".rds")
download.file(url, destfile = destfile)
packages <- as.data.frame(readRDS(destfile), stringsAsFactors = FALSE)
sum(packages$Package == "arrow")
```

giving

```
> sum(packages$Package == "arrow")
[1] 1
```

Because that version of the package is not declared in the repository "manifest", and because it's not located in the typical Archive path, `renv` is unable to find it. In theory, `renv` could "guess" and look for the package in that location, but I think the repository PACKAGES file should be fixed.

**Comment by vspinu** [comment]
Ok, I have reported it upstream. 

One thing is not very clear to me, PACKAGES.rds does not contain paths, so I would assume `https://arrow-r-nightly.s3.amazonaws.com/src/contrib/<name>_X.Y.Z` is a typical path because `install.pcakges` does figure out the only available package there? Or the path is retrieved from other metadata location?

**Comment by kevinushey** [comment]
My understanding is that `install.packages()` uses the PACKAGES manifest to find the "latest" available declared package; because arrow 6.0.0 is registered there but not arrow 5.0.0, arrow 6.0.0 is chosen to be installed. So `install.packages()` here can find arrow 6.0.0, but not arrow 5.0.0.

It's a little confusing because, as far as I know, there isn't any explicit documentation for the structure of the PACKAGES manifest file. However, entries within the PACKAGES file can include a Path field; for example:

```
Package: renv
Version: 1.0.0
License: GPL
MD5sum: dd01b2d7a4575cf20ee08dc79061819f
NeedsCompilation: no
Path: 4.1.0/Recommended
```

In this case, the package repository would contain the `renv` archive at `src/contrib/4.1.0/Recommended/renv_1.0.0.tar.gz`. This is most commonly used by CRAN when preparing for a new release of R, where packages build for the next release of R are stored in a separate sub-directory and those entries are then added to the PACKAGES file.

**Comment by vspinu** [comment]
I see. This makes sense. When `Path` is missing, it default to empty string and thus the location would be `src/contrib/arrow_5.0.0.20211025.tar.gz`, exactly what is needed in this case. I guess this is how `install.packages` finds the only available package then.
